src/m3ninx/search/executor/iterator.go

```diff
@@ -79,12 +88,24 @@ func (it *iterator) Next() bool {
 			return false
 		}
 		it.totalSearchDuration = time.Since(start)
+		if !it.next() {
+			it.done = true
+			return false
+		}
+		it.nextDoc = it.current()
 	}
-	if it.idx == len(it.iters) {
-		return false
+
+	it.currDoc = it.nextDoc
```

Comment:
i went with this look-ahead impl instead of propagating the `HasNext` to all the various iterators (doc, postingslist, roaring, etc) in the query path.